### ISSUE TITLE
offset option accepts junk arguments with no validation error

### ISSUE THREAD

**Issue by brchrisman** [axboe/fio#841]
I can pass --offset=maryhadalittlelamb with no non-zero exit status or output on stderr.
Verified on version fio-3.16-15-g983a65

Not very familiar with the code, I see this looks like the option parser

> 	{
> 		.name	= "offset",
> 		.lname	= "IO offset",
> 		.alias	= "fileoffset",
> 		.type	= FIO_OPT_STR_VAL,
> 		.cb	= str_offset_cb,

Calling 

> static int str_offset_cb(void *data, unsigned long long *__val)

which gets through eventually to

> parse_is_percent

... and presumably is somehow taking the string passed in and checking whether it's between 0 and 100 as though the user is passing in a unsigned long long on the command-line?

static inline int parse_is_percent(unsigned long long val)

> 	return val <= -1ULL && val >= (-1ULL - 100ULL);

the plumbing to get there seems to go straight through
		fio_opt_str_val_fn *fn = o->cb;

and the cb seems to lead straight to the apparently blind check of the option as though it's an ULL?

It's not immediately clear to me where/how this would be fixed (such that fio exits with an error if the --offset option doesn't parse properly).  If there's a suggestion on how to do that, I might be able to put a patch together.

**Comment by axboe** [comment]
For conversions like this, fio basically relies on strtoll(3) from libc to do the conversion. If it doesn't return an error, then we trust what it returns. No further validation happens of the string.

**Comment by axboe** [comment]
Pushed an improvement for this, please test current -git.

**Comment by brchrisman** [comment]
Hey thanks, much appreciated!

> [build@auto-worker-0000005251 fio]$ !526
> ./fio --warnings-fatal --thread --offset=bahbahblacksheep$((1<<20))--direct=1 --runtime=$runtime                         --time_based --output-format=json --filename=$device --name ${MYCLUSTER_IP}_${testUUID} --rw=$pattern --bs=$size --iodepth=bibly --size=$((1<<24)) > /dev/null
> fio: failed parsing offset=bahbahblacksheep1048576--direct=1
> [build@auto-worker-0000005251 fio]$ echo $?
> 1
> [build@auto-worker-0000005251 fio]$

**Comment by axboe** [comment]
You're welcome! Note that it's still possible to trick it, for instance if you do:

```
offset=0xbahbahblacksheep
```

it'd pass, since the start of the string was valid. But I think at that point you're deliberately trying to trick it, and there's little point in trying to catch that as well. At least now we catch inadvertent errors, which is the most important imho.

**Comment by brchrisman** [comment]
The original problem which got me looking at it, was a screw up in my scripting where I submitted: 

> --offset=1048576--direct=1 

and my test runs ran without O_DIRECT and no error.  As you mention, I suppose this doesn't directly address that, but it's a big improvement.  I focused on the nasty example because I was trying to chase the code down.
I didn't catch that ret = check_str_bytes() was returning a value-for-use.... so I used the 'mary had a little lamb' example to try and chase the bug down.
